lib/explorer/data_frame.ex

```diff
@@ -104,39 +104,139 @@ defmodule Explorer.DataFrame do
+  defguard is_column_name(column) when is_binary(column) or is_atom(column)
+  defguard is_column_pairs(columns) when is_list(columns) or is_map(columns)
+
+  # Normalize a column name to string
+  defp to_column_name(column) when is_binary(column), do: column
+  defp to_column_name(column) when is_atom(column), do: Atom.to_string(column)
+
+  # Normalize pairs of `{column, value}` where value can be anything.
+  # The `column` is only validated if it's an integer. We check that the index is present. 
+  defp to_column_pairs(df, pairs), do: to_column_pairs(df, pairs, & &1)
+
+  # The function allows to change the `value` for each pair.
+  defp to_column_pairs(df, pairs, value_fun)
+       when is_column_pairs(pairs) and is_function(value_fun, 1) do
+    existing_columns = names(df)
+
+    pairs
+    |> Enum.map_reduce(nil, fn
+      {column, value}, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column = fetch_column_at!(map, column)
+
+        {{existing_column, value_fun.(value)}, map}
+
+      {column, value}, maybe_map when is_atom(column) ->
+        column = Atom.to_string(column)
+
+        {{column, value_fun.(value)}, maybe_map}
+
+      {column, value}, maybe_map when is_binary(column) ->
+        {{column, value_fun.(value)}, maybe_map}
+    end)
+    |> then(fn {pairs, _} -> pairs end)
+  end
+
+  defp fetch_column_at!(map, column) do
+    case Map.fetch(map, column) do
+      {:ok, column} -> column
+      :error -> raise ArgumentError, "column does not exist at index #{column}"
+    end
+  end
+
+  defp column_index_map(names),
+    do: for({name, idx} <- Enum.with_index(names), into: %{}, do: {idx, name})
+
+  # Normalize column names without verifying if they exist.
+  defp to_column_names(names) when is_list(names),
+    do: Enum.map(names, &to_column_name/1)
+
+  # Normalize column names and raise if column does not exist.
+  defp to_existing_columns(df, columns) when is_list(columns) do
+    existing_columns = names(df)
+
+    columns
+    |> Enum.map_reduce(nil, fn
+      column, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column = fetch_column_at!(map, column)
+
+        {existing_column, map}
+
+      column, maybe_map when is_atom(column) ->
+        column = Atom.to_string(column)
+        maybe_raise_column_not_found(existing_columns, column)
+        {column, maybe_map}
+
+      column, maybe_map when is_binary(column) ->
+        maybe_raise_column_not_found(existing_columns, column)
+        {column, maybe_map}
+    end)
+    |> then(fn {columns, _} -> columns end)
+  end
+
+  defp to_existing_columns(df, %Range{} = columns) do
+    existing_columns = names(df)
+
+    Enum.map(columns, fn idx ->
+      case Enum.at(existing_columns, idx) do
+        nil -> raise ArgumentError, "column does not exist at index #{idx}"
+        column -> column
+      end
+    end)
+  end
```

Comment:
```suggestion
  defp to_existing_columns(df, %Range{} = columns) do
    Enum.slice(names(df), columns)
  end
```

And this will actually put us half way to supporting #204.